projects/WTI-API/src/main/services/ScoreboardChangeListener.java

```diff
@@ -0,0 +1,170 @@
+	public void runDeleted(IRun run) {
+		markStandingsNotCurrent();
+	}
+
+	/**
+	 * This method does nothing, since checking out a run (to judge it) has no effect on standings.
+	 */
+	@Override
+	public void runCheckedOut(IRun run, boolean isFinal) {
+		// Nothing; this can't change the standings
+	}
+
+	/**
+	 * Sets the contest controller's "wtiServerStandingsAreCurrent" flag false, since a
+	 * judged run potentially changes the standings.
+	 */
+	@Override
+	public void runJudged(IRun run, boolean isFinal) {
+		markStandingsNotCurrent();
+	}
+
+	/**
+	 * Sets the contest controller's "wtiServerStandingsAreCurrent" flag false, since an
+	 * updated run potentially changes the standings.
+	 */
+	@Override
+	public void runUpdated(IRun run, boolean isFinal) {
+		markStandingsNotCurrent();
+	}
+
+	/**
+	 * This method does nothing, since compiling a run has no effect on standings.
+	 */
+	@Override
+	public void runCompiling(IRun run, boolean isFinal) {
+		// Nothing; this can't change the standings
+	}
+
+	/**
+	 * This method does nothing, since executing a run has no effect on standings.
+	 */
+	@Override
+	public void runExecuting(IRun run, boolean isFinal) {
+		// Nothing; this can't change the standings
+	}
+
+	/**
+	 * This method does nothing, since validating a run has no effect on standings.
+	 */
+	@Override
+	public void runValidating(IRun run, boolean isFinal) {
+		// Nothing; this can't change the standings
+	}
+
+	/**
+	 * This method does nothing, since cancelling the judging of a run has no effect on standings.
+	 */
+	@Override
+	public void runJudgingCanceled(IRun run, boolean isFinal) {
+		// Nothing; this can't change the standings
+	}
+
+	/**
+	 * Sets the contest controller's "wtiServerStandingsAreCurrent" flag false, since
+	 * adding a configuration item (for example, a new team account) potentially changes the standings.
+	 */
+	@Override
+	public void configurationItemAdded(ContestEvent contestEvent) {
+		markStandingsNotCurrent();
+	}
+
+	/**
+	 * Sets the contest controller's "wtiServerStandingsAreCurrent" flag false, since
+	 * updating a configuration item (for example, changing the "display on scoreboard" property of
+	 * a team acount) potentially changes the standings.
+	 */
+	@Override
+	public void configurationItemUpdated(ContestEvent contestEvent) {
+		markStandingsNotCurrent();
+	}
+
+	/**
+	 * Sets the contest controller's "wtiServerStandingsAreCurrent" flag false, since
+	 * removing a configuration item (for example, a team account) potentially changes the standings.
+	 */
+	@Override
+	public void configurationItemRemoved(ContestEvent contestEvent) {
+		markStandingsNotCurrent();
+	}
+	
+	/**
+	 * Sets "wtiServerStandingsAreCurrent" in ContestController to false; 
+	 * sends a websocket message to each team client indicating standings are not current.
+	 */
+	private void markStandingsNotCurrent() {
+		
+		//tell the ContestController that its cached standings should no longer be considered current

```

Comment:
I cannot see any reason that this method needs to be synchronized.  I base this on the following analysis:

If multiple threads execute this method simultaneously, each thread basically encounters four statements in some unknown and potentially interleaved order:

1. `contestController.setWtiServerStandingsAreCurrent(false) ;`

It's not going to matter if multiple threads execute this statement in parallel; the result is always going to be the same: the "standings are current" flag in the controller is going to be (repeatedly) set to false.

2.  `HashMap<String, ServerConnection> teamConnections = contestController.getTeamConnections();`

It's true that different threads could get different lists of teamConnections (if a new team is added during thread executions or if a team is dropped), but that could happen even if the method was synchronized (in fact, it's MORE LIKELY to happen if the method is synchronized, because there would be more time for the team list to change while threads are blocked on the synchronized entry), and anyway it doesn't seem that it would make any difference: each thread would have its own list of teams to which it sends messages and synchronizing might change the team list but this shouldn't matter -- different threads can get different team lists with or without synchronizing the method.

Then there is a for-loop `for(each team)`, which does two things on each interation (i.e., for each team in its list):

3.  `			JsonObject builder = Json.createObjectBuilder()
				.add("type", WebsocketMsgType.STANDINGS.name().toLowerCase())
				.add("id", "1-1") 
				.add("teamId", teamkey)
				.build();
`

Each thread is going to build its own JSON Object containing the current team's "teamkey", and two threads doing this in parallel are still going to produce the correct result in each thread.

4.  `socket.sendMessage(builder.toString());`

The `socket` object here is a `WTIWebsocket`.  The `sendMessage()` method in that class has one statement:  `this.session.getAsyncRemote().sendText(message);`.  The `session` object in a `WTIWebsocket` is a `javax.websocket.Session` object, whose documentation states:
`Session objects may be called by multiple threads. Implementations must ensure the integrity of the mutable properties of the session under such circumstances.`

Based on this, there does not seem to be any reason that two threads can't simultaneously send a message to a given socket.  In this case the message will be the same from each thread:  a `WebsocketMsgType.STANDINGS` message.  This is an indication to clients that the Standings are no longer current, and the message goes separately to each team connection -- it does not matter in what order the teamConnections get the message.

Taken together, the above analysis seems to indicate that there's no _need_ to synchronize the method. 
Since adding `synchronized` to the method would potentially slow things down and make things less scalable, I'm suggesting we don't do that.

If you have a different analysis or counter-argument that indicates that the method should indeed be synchronized, please present it.

It _**is**_ potentially arguable that the last statement (`socket.sendMessage(builder.toString());`) should be done on a separate thread for each message send -- as noted in the related PR comment -- but that doesn't seem to indicate any reason that the entire method should be synchronized.

Reply:
Since Doug has approved the PR, I'm assuming he accepts the argument presented as to why method `markStandingsNotCurrent()` does _not_ need to be synchronized, so I'm marking this conversation "resolved".